[visionOS] Audio from YouTube video is not spatial when multiple Safari windows are open.

mwyrzykowski opened the pull request:
#### ffbdce19ceb5cd4087827fa01bc39794ec1823ec
<pre>
[visionOS] Audio from YouTube video is not spatial when multiple Safari windows are open.
<a href="https://bugs.webkit.org/show_bug.cgi?id=261609">https://bugs.webkit.org/show_bug.cgi?id=261609</a>
&lt;radar://115086235&gt;

Reviewed by Dean Jackson.

Pass the sceneIdentifier down from the UI process to the GPU process so
that the AVAudioSession can use it to spatialize from the correct window.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::visibilityStateChanged):

* Source/WebCore/page/Page.cpp:
(WebCore::Page::sceneIdentifier const):
(WebCore::Page::setSceneIdentifier):
Store the scene identifier.

* Source/WebCore/page/Page.h:
Store the sceneIdentifier the Page corresponds to.

* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::setPageIsVisible):
Pass the sceneIdentifier when changing the visibility of the MediaPlayer.

* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.cpp:
(WebCore::MediaPlayerPrivateAVFoundation::setPageIsVisible):
Pass the sceneIdentifier when the visibility changes.

* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::setPageIsVisible):
Call setIntendedSpatialExperience with a corresponding sceneIdentifier.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::setPageIsVisible):
sceneIdentifier is not needed for MediaPlayerPrivateMediaStreamAVFObjC

* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::setPageIsVisible):
Or the webM player.

* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp:
(WebCore::MockMediaPlayerMediaSource::setPageIsVisible):
Member function is a no-op.

* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::setPageIsVisible):
Pass the sceneIdentifier along.

* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
Pass the sceneIdentifier along with SetPageIsVisible.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _registerForNotifications]):
(-[WKWebView _webViewWindowDidBecomeKey:]):
(-[WKWebView _webViewSceneWillEnterBackground:]):
(-[WKWebView _webViewSceneWillEnterForeground:]):
(-[WKWebView _webViewSceneDidActivate:]):
Update the sceneIdentifier whenever we receive a scene notification.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::setSceneIdentifier):
Pass the sceneIdentifier from the UI to the web process.

* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::setPageIsVisible):
Pass the sceneIdentifier from the web to GPU process.

* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setSceneIdentifier):
Update the WebCore::Page&apos;s sceneIdentifier.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Add a message to pass the sceneIdentifier.

Canonical link: <a href="https://commits.webkit.org/268152@main">https://commits.webkit.org/268152@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4af4a62a23802d5787f41d76ad0d11f83388f594

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18828 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19171 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19775 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20696 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17625 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19022 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22476 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19309 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19422 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19054 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19190 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16402 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21579 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16410 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17163 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23599 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17439 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17335 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21510 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17931 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15209 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16994 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4480 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21361 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17767 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -221,8 +221,39 @@ - (void)_registerForNotifications
+    [center addObserver:self selector:@selector(_webViewSceneWillEnterForeground:) name:UISceneWillEnterForegroundNotification object:nil];
+    [center addObserver:self selector:@selector(_webViewSceneDidActivate:) name:UISceneDidActivateNotification object:nil];
+    [center addObserver:self selector:@selector(_webViewSceneWillEnterBackground:) name:UISceneDidEnterBackgroundNotification object:nil];
+    [center addObserver:self.window selector:@selector(_webViewWindowDidBecomeKey:) name:UIWindowDidBecomeKeyNotification object:nil];
```
`addObserver:self.window` seems incorrect – shouldn't it be `addObserver:self`?

Also, should we use a non-nil object (either the scene or window), depending on the notification, so that we're not listening for notifications from objects we don't care about?

mwyrzykowski replied:
Oh yes it should be self and good idea, I will use a non-nil object

mwyrzykowski replied:
Actually I'm going to remove the notifications. They don't occur at the times I need and tracking the state and setting it later seems unnecessary.

rr-codes commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
```diff
@@ -429,11 +429,29 @@ void getSupportedTypes(HashSet<String>& types) const final
+    AVAudioSession *session = [PAL::getAVAudioSessionClass() sharedInstance];
     if (!visible) {
-        AVAudioSession *session = [PAL::getAVAudioSessionClass() sharedInstance];
+        if (NSString *sceneId = sceneIdentifier; sceneId.length > 0) {
```
the `> 0` should omitted, right?

grorg commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -7383,7 +7389,12 @@ void HTMLMediaElement::createMediaPlayer() WTF_IGNORES_THREAD_SAFETY_ANALYSIS
     m_player->setMuted(effectiveMuted());
+#if PLATFORM(IOS_FAMILY)
+    auto page = document().page();
+    m_player->setPageIsVisible(!m_elementIsHidden, page ? page->sceneIdentifier() : ""_s);
```
I have no opinion here, but I wonder if it would be better to have the `#if` conditional inside `sceneIdentifier()` rather than the calling site. i.e. `Page::sceneIdentifier` is on all platforms, and returns `""_s` everywhere but IOS_FAMILY.

mwyrzykowski replied:
That seems like a better idea, I'll make that change

mwyrzykowski committed:
[visionOS] Audio from YouTube video is not spatial when multiple Safari windows are open.
https://bugs.webkit.org/show_bug.cgi?id=261609
<radar://115086235>

Reviewed by Dean Jackson.

Pass the sceneIdentifier down from the UI process to the GPU process so
that the AVAudioSession can use it to spatialize from the correct window.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::visibilityStateChanged):

* Source/WebCore/page/Page.cpp:
(WebCore::Page::sceneIdentifier const):
(WebCore::Page::setSceneIdentifier):
Store the scene identifier.

* Source/WebCore/page/Page.h:
Store the sceneIdentifier the Page corresponds to.

* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::setPageIsVisible):
Pass the sceneIdentifier when changing the visibility of the MediaPlayer.

* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.cpp:
(WebCore::MediaPlayerPrivateAVFoundation::setPageIsVisible):
Pass the sceneIdentifier when the visibility changes.

* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::setPageIsVisible):
Call setIntendedSpatialExperience with a corresponding sceneIdentifier.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::setPageIsVisible):
sceneIdentifier is not needed for MediaPlayerPrivateMediaStreamAVFObjC

* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.h:
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::setPageIsVisible):
Or the webM player.

* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.cpp:
(WebCore::MockMediaPlayerMediaSource::setPageIsVisible):
Member function is a no-op.

* Source/WebCore/platform/mock/mediasource/MockMediaPlayerMediaSource.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::setPageIsVisible):
Pass the sceneIdentifier along.

* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
Pass the sceneIdentifier along with SetPageIsVisible.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _registerForNotifications]):
(-[WKWebView _webViewWindowDidBecomeKey:]):
(-[WKWebView _webViewSceneWillEnterBackground:]):
(-[WKWebView _webViewSceneWillEnterForeground:]):
(-[WKWebView _webViewSceneDidActivate:]):
Update the sceneIdentifier whenever we receive a scene notification.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::setSceneIdentifier):
Pass the sceneIdentifier from the UI to the web process.

* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::setPageIsVisible):
Pass the sceneIdentifier from the web to GPU process.

* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setSceneIdentifier):
Update the WebCore::Page's sceneIdentifier.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Add a message to pass the sceneIdentifier.

Canonical link: https://commits.webkit.org/268152@main
